File: python/72.py
class Solution(object):
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        w1,w2,opened=word1,word2,set()
        topens=[(w1,w2,0)]
        while True:
            (w1,w2,value)=topens.pop(0)
            if (w1,w2) in opened:
                continue
            if w1==w2:
                return value
            opened.add((w1,w2))
            while w1 and w2 and w1[0]==w2[0]:
                w1=w1[1:]
                w2=w2[1:]
            value=value+1
            topens+=[(w1[1:],w2,value),(w1,w2[1:],value),(w1[1:],w2[1:],value)]
        return -1
            

File: java/dp_2/72.java
class Solution {
    public int minDistance(String word1, String word2) {
        int s = word1.length();
        int l = word2.length();
        int[][] arr = new int[s + 1][l + 1];
        for (int i = 0; i <= s; i++) {
            arr[i][0] = i;
        }
        for (int j = 0; j <= l; j++) {
            arr[0][j] = j;
        }
        for (int i = 0; i < s; i++) {
            for (int j = 0; j < l; j++) {
                if (word1.charAt(i) == word2.charAt(j)) {
                    arr[i + 1][j + 1] = arr[i][j];

                } else {
                    arr[i + 1][j + 1] = Math.min(Math.min(arr[i + 1][j], arr[i][j + 1]), arr[i][j]) + 1;

                }
            }
        }
        return arr[s][l];

    }
}